app/client/keybase/README.md

```diff
@@ -0,0 +1,191 @@
+
+
+### Makefile Helper
+
+
+To aid in the testing of the local keybase the following `Makefile` command has been exposed `make test_app` which will run the test suites from the `app` module alone, which includes the `app/client/keybase/keybase_test.go` file which covers the functionality of the `Keybase` implementation
+
+
+## KeyPair Interface
+
+
+The `KeyPair` interface exposes methods related to the operations used on the pairs of `PublicKey` types and JSON encoded, `PrivKeyArmour` strings.
+
+```go
+// The KeyPair interface exposes functions relating to public and encrypted private key pairs
+type KeyPair interface {
+	// Accessors
+	GetPublicKey() PublicKey
+	GetPrivArmour() string
+
+	// Public Key Address
+	GetAddressBytes() []byte
+	GetAddressString() string // hex string
+
+	// Unarmour
+	Unarmour(passphrase string) (PrivateKey, error)
+
+	// Export
+	ExportString(passphrase string) (string, error)
+	ExportJSON(passphrase string) (string, error)
+}
+```
+
+The `KeyPair` interface is implemented by the `EncKeyPair` struct
+
+```go
+// EncKeyPair struct stores the public key and the passphrase encrypted private key
+type EncKeyPair struct {
+	PublicKey     PublicKey
+	PrivKeyArmour string
+}
+```
+
+The `EncKeyPair` struct stores the `PublicKey` of the key pair and the JSON encoded, `"armoured"` key string. The armoured key string is built from the following struct
+
+```go
+// Armoured Private Key struct with fields to unarmour it later
+type ArmouredKey struct {
+	Kdf        string `json:"kdf"`
+	Salt       string `json:"salt"`
+	SecParam   string `json:"secparam"`
+	Hint       string `json:"hint"`
+	CipherText string `json:"ciphertext"`
+}
+```
+
+This struct is created after the [encryption step](#encryption-and-armouring) has encrypted the private key string into the `CipherText` and the other fields are filled using the parameters used in the [encryption step](#encryption-and-armouring) so that the armoured key can later be unarmoured and decrypted provided the correct passphrase is supplied. This struct is then marshalled into a JSON string and stored in the `EncKeyPair`'s `PrivKeyArmour` field.
+
+
+## KeyPair Code Structure
+
+The KeyPair code is seperated into two files `shared/crypto/keypair.go` and `shared/crypto/armour.go`
+
+```
+[ 146] shared
+└──── [ 102] crypto
+      ├──── [ 5KB] armour.go
+      └──── [ 3KB] keypair.go
+```
+
+
+## Encryption and Armouring
+
+Whenever a new key is created or imported it is encrypted using the passphrase provided (this can be `""` for no passphrase).
+
+The packages used for the encryption and armouring of a key are: `golang.org/x/crypto/scrypt` (key generation), `crypto/aes`, `crypto/cipher` (AES256-GCM encryption cipher) as well as `encoding/hex`, `encoding/base64` and `encoding/json` for the armouring of the encrypted key.
+
+1. First the OS's randomness is used to generate a `[]byte` of length `16` which is used as the salt for generating the key.
```

Comment:
It feels like there's too much detail here.

The details, libraries, functions, etc made me want to go and read through the code rather than the documentation.

The following isn't complete but I wanted to put together something simple as a starting point that would give a general idea of what how the implementation works:

Since this is just pseudocode, note that we have the freedom to play around where/when we add a certain amount of detail.

```mermaid
flowchart LR
    subgraph C[core lib]
        A["rand([16]byte)"]
    end
    subgraph S[scrypt lib]
        B["key(salt, pass, ...)"]
    end
    subgraph AES-GCM
        direction TB
        D["Cipher(key)"]
        E["GCM(block)"]
        F["Seal(plaintext, nonce)"]
        D--Block-->E
        E--Nonce-->F
    end
    C--Salt-->S
    S--Key-->AES-GCM
```

![Screenshot 2023-01-27 at 8 07 06 PM](https://user-images.githubusercontent.com/1892194/215233047-08b0d751-a901-4f21-8b77-d5c382288d6e.png)